### DISCUSSION TITLE
Doubt on comments on API.cmake

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#1194] [Q&A] [answered]
Why in the example of the comments at the lines https://github.com/nasa/fprime/blob/85fa25bc0dd211efefd184b87db95f774ea687c5/cmake/API.cmake#L80-L91 and at the lines https://github.com/nasa/fprime/blob/85fa25bc0dd211efefd184b87db95f774ea687c5/cmake/API.cmake#L191-L201  LINK_DEPS is used in the snippet of code? The comment is not about the variable MOD_DEPS?

**Comment by LeStarch** [comment, marked as answer]
This looks like an oversight.  I am going to make this into an issue and we'll get it resolved.

**Comment by SMorettini** [comment]
About those variables, there is something else I don't understand.
When I was using FPrime 1.5.3 my CMakeList.txt had these lines for using external libraries:
```
find_package (orocos_kdl REQUIRED)
include_directories (${orocos_kdl_INCLUDE_DIRS})

find_package(OpenCV REQUIRED)
include_directories (${OpenCV_INCLUDE_DIRS})

...

set(MOD_DEPS
    "${orocos_kdl_LIBRARIES}"
     "${OpenCV_LIBRARIES}"
)

...

set(UT_MOD_DEPS
    "${orocos_kdl_LIBRARIES}"
    "${OpenCV_LIBRARIES}"
)

```
After migrating to 2.0.1 using the previous CMakeList.txt I have the following error:
```
No rule to make target 'MyAwesomeModule/opencv_videoio_package_gen', needed by 'MyAwesomeModule/CMakeFiles/MyAwesomeModule_package_gen'.  Stop.
No rule to make target 'MyAwesomeModule/opencv_videoio_package_gen', needed by 'MyAwesomeModule/CMakeFiles/MyAwesomeModule_package_gen'.  Stop.
```
I tried multiple solutions and I managed to make it works using these lines:
```
find_package (orocos_kdl REQUIRED)
include_directories (${orocos_kdl_INCLUDE_DIRS})

find_package(OpenCV REQUIRED)
include_directories (${OpenCV_INCLUDE_DIRS})

...

set(MOD_DEPS
    "-l${orocos_kdl_LIBRARIES}" #### added the -l ####
)
set(LINK_DEPS
    "${OpenCV_LIBRARIES}"    ###Only adding the -l as with orocos_kdl was giving the error as above
)
...

set(UT_MOD_DEPS
    "${orocos_kdl_LIBRARIES}"
    "${OpenCV_LIBRARIES}"
)

```
The questions are:

1. Am I using the right approach to use external libraries?
2. If I'm using the right approach, why for `orocos_kdl` adding the `-l` works and for OpenCV no? 
3. Why for `UT_MOD_DEPS` I don't need to add `-l`(as it was for the FPrime 1.5.3)? 

p.s.: I was not sure If I had to continue the discussion here or open a new one. If I have to open a new one tell me and I'll move this post to a new discussion.

**Reply by LeStarch** [reply]
Some help:

1. Remove quotes.  I think your libraries are lists, as are mod deps.  If you use quotes you'' get an internally escaped list (e.g. "orocoslib1;orocoslib2" "OpenCVlib1;OpenCVlib2"
```
set(MOD_DEPS
     ${orocos_kdl_LIBRARIES}
     ${OpenCV_LIBRARIES}
)

```
2. To fix `MyAwesomeModule/opencv_videoio_package_gen`, you can add dummy targets.
```
add_custom_target(MyAwesomeModule/opencv_videoio_package_gen)
```
This is a known issue with F´ < v3.0.0.

**Reply by LeStarch** [reply]
Hmmm, you have a point `-l` is needed for non-F´ libraries.  Here is likely the cleanest work around

```
find_package (orocos_kdl REQUIRED)
include_directories (${orocos_kdl_INCLUDE_DIRS})

find_package(OpenCV REQUIRED)
include_directories (${OpenCV_INCLUDE_DIRS})
...
...
add_fprime_module()
# Code comes after add F´ module
get_module_name()
target_link_libraries("${MODULE_NAME}"  ${orocos_kdl_LIBRARIES} ${OpenCV_LIBRARIES})
...
... 
```
Note: here we are not setting MOD_DEPS at all, but just adding them as link libraries to the module directly.  Note: you should only call target_link_libraries after add_fprime_module as it creates the given target.

**Reply by LeStarch** [reply]
You could also add -l to each library and then set it in mod deps....but that is more painful.
